### ISSUE TITLE
Link does not support path in Unicode character

### ISSUE THREAD

**Issue by johnking** [gatsbyjs/gatsby#10058]
## Description

I created a website with Gatsby and found an issue, the `<Link>` component does not support unicode path like:   ` <Link to="首页">首页</Link>`

I also tried the `Link` component directly from` @reach/router` package and found the same issue, so I submitted an issue over there. https://github.com/reach/router/issues/206

To help other people with the same issue, I am adding the issue here. 

### Steps to reproduce

Please take a look at the running example at codesandbox: 

https://codesandbox.io/s/l5vl5xr0nq

### Expected result

after clicking the `<Link to="首页">首页</Link>`

it should jump to page `首页`.

### Actual result

it does not handle the navigation request. 

### Environment

https://codesandbox.io/s/l5vl5xr0nq

**Comment by pieh** [comment]
Thanks for creating issue and tracking it down to bug (or limitation) in `@reach/router`. We will have to see first how the issue will be handled there to see what's best course of action is here. If this won't be fixed upstream we will need to display warnings when creating pages that contain unicode characters in path

**Comment by alaycock** [comment]
The current way you would handle this in @reach/router would be to use `encodeURI`. For example:
```
  <Router>
    <Home path="/" />
    <Dashboard path="dashboard" />
    <Unicode path={encodeURI("首页")} />
  </Router>
```
I haven't used Gatsby, but would you be able to wrap each path in `encodeURI` when defining the router? It doesn't escape any of: `    A-Z a-z 0-9 ; , / ? : @ & = + $ - _ . ! ~ * ' ( ) #`, so it shouldn't affect most english-language routes, although, I don't know how it would affect anyone else who is currently using unicode Gatsby.

https://codesandbox.io/s/mn8jn27mj

**Comment by johnking** [comment]
Hi @alaycock , 

This workaround works only if we create the routes programmatically, but as said in my previous feedback at  https://github.com/reach/router/issues/206, it won't work for static routes like the example below.

```yml
- title: 般若摄颂第二十课
  items:
    - id: si-kao-ti
      title: 思考题
      href: /scripts/bo-re-she-song-20.html#思考题
      forceInternal: true
    - id: 丑二、获得报身之功德
      title: 丑二、获得报身之功德
      href: /scripts/bo-re-she-song-20.html#丑二、获得报身之功德
      forceInternal: true
 ...
```

For example, we don't know how to process the links like `href: /scripts/bo-re-she-song-20.html#思考题` with this workaround.

Even for large open source project like [react documentation](https://reactjs.org/), it could not use the existing code to display the translated version (Chinese) of docs because all the hrefs in all `yaml` files are predefined.

  So we have to support route path in unicode unless we have other workarounds.

thanks

-John

**Comment by johnking** [comment]
@alaycock 

I have tried your workaround with my application, and it did not work as expected.

What I did: 

1. bypass `Gatsby <link> ` and directly call `@reach/router <link>` 
2. wrap all route paths with `encodeURI()`

I am posting a screenshot and you might find all the links are encodeURI-ed in chrome devtool console.

<img width="1252" alt="screen shot 2018-11-24 at 12 48 02" src="https://user-images.githubusercontent.com/1787137/48971422-7721fc80-efe7-11e8-8926-72ed14455699.png">

Basically, the client routing does not work but **the server side routing still works, it will jump to the right section by refresh the page.**

thanks

-John

**Comment by wendzhue** [comment]
Ooh. This @gatsbot did a great job to pretend that the problem is fixed. :(

**Comment by durunvo** [comment]
This would be really helpful

**Comment by stahldom** [comment]
Still an ongoing issue. Not fixed.
